### PR TITLE
metamorphic: add IngestAndExcise, make EFOS determinitic

### LINKED ISSUE #2885: metamorphic: integrate IngestWithExcise
The excise operation is currently untested by the metamorphic tests. Integration could've caught issues like the one fixed in #2884.

### PR DESCRIPTION
**Pull request by itsbilal**
This change updates the metamorphic tests to support a new useExcise testOption and a new IngestAndExcise operation, which calls db.IngestAndExcise() if useExcise is true, or calls it with an sstable containing DeleteRange/RangeKeyDelete for the same span if useExcise is false.

Furthermore, a new private option is added, efosAlwaysCreatesIterators. If the test option for useExcise (or the previously added useSharedReplicate) is enabled, one of these two things happen to guarantee EFOS determinism around excises:

1) If efosAlwaysCreatesIterators is true, the behaviour of EFOS is changed to always hold onto a readState and never return errors from NewIter.
2) If efosAlwaysCreatesIterators is false, regular EFOS behaviour is retained, but a flush is completed after every EFOS instantiation to guarantee that it becomes excise-immune right at instantiation.

Note that regular snapshots are never created when useExcise or useSharedReplicate is true.

Fixes #2885.

### CODE REVIEW

**Comment by itsbilal** [comment]
This isn't quite ready to merge yet (the body of `ingestAndExciseOp` is a little rough right now), but I sped it up to help investigate https://github.com/cockroachdb/cockroach/issues/114056 and https://github.com/cockroachdb/cockroach/issues/113973 , and sure enough it fails pretty readily when run locally.

**Comment by itsbilal** [comment]
There's one occasional bug/failure around virtual sstable fileMetadata refcounts that this test is still tripping up. I'll look into that momentarily, but that bug aside (which is likely not in the test and in pebble code itself), this is ready for a review.

**Comment by itsbilal** [comment]
Note that only the last commit needs a review. The first couple commits are #3074.

**Comment by itsbilal** [comment]
Rebased on master so it should just have the one commit for the metamorphic test changes now.

**Review by jbowens — COMMENTED** [review]
Looks like there's still a metamorphic test failure in CI?

```
===== SEED =====

[41](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:42) 1700494649026393988

[42](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:43) ===== DIFF =====

[43](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:44) \_meta/231120-153729.0263362307724/{standard-000,standard-004}

[44](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:45) @@ -1595,12 +1595,12 @@

[45](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:46) db1.Get("dojyrmo") // \[""\] pebble: not found #1594

[46](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:47) iter17.First() // \[true,"anawodd",<no point>,\["anawodd","anawodd@28")=>{""="pvpltzv","@15"="iefqtmkpw"}\*\] <nil> #1595

[47](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:48) db1.Get("anawodd@28") // \[""\] pebble: not found #1596

[48](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:49) iter15.Last() // \[true,"dojyrmo@50","ymtjebhutgk",<no range>\] <nil> #1597

[49](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:50) iter15.Prev("") // \[false\] <nil> #1598

[50](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:51)\-iter17.NextPrefix() // \[true,"dojyrmo@50","ymtjebhutgk",<no range>\*\] <nil> #1599

[51](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:52)\-iter17.Last() // \[true,"zyolnpsqklu@2","dwl",<no range>\] <nil> #1600

[52](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:53)+iter17.NextPrefix() // \[true,"anawodd@28",<no point>,\["anawodd@28","bifqyolqz@45")=>{""="pvpltzv"}\*\] <nil> #1599

[53](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:54)+iter17.Last() // \[true,"zyolnpsqklu@2","dwl",<no range>\*\] <nil> #1600

[54](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:55) iter17.Last() // \[true,"zyolnpsqklu@2","dwl",<no range>\] <nil> #1601

[55](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:56) iter17.SeekLT("dojyrmo@24", "") // \[true,"dojyrmo@34","jieppgnxeabys",<no range>\] <nil> #1602

[56](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:57) iter17.Next("") // \[true,"dojyrmo@15","qjsanhnsxg",<no range>\] <nil> #1603

[57](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:58) iter16.Next("") // \[false\] <nil> #1604

[58](https://github.com/cockroachdb/pebble/actions/runs/6932349203/job/18856077586?pr=3044#step:4:59) iter15.Next("") // \[true,"dojyrmo@50","ymtjebhutgk",<no range>\] <nil> #1605
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3044#-:-NjsRKmzEd6czwuMnWEP:bq6klcl)* status: 0 of 18 files reviewed, all discussions resolved

**Comment by itsbilal** [comment]
@jbowens yep, latest push fixes it. basically in the non-shared replicate case we always did a DeleteRange+RangeKeyDelete but in the shared replace case we would do an excise only if there's at least one shared or one unshared file. I've addressed it now.

The boomerang-file issue was actually ^ so all known issues should be fixed now.

**Comment by itsbilal** [comment]
I'll take a look at that failure. We should still review and aim to merge this soon though, as this test has already found multiple genuine bugs (including one that KV had to triage) and new failures are far more likely to be bugs in Pebble itself than in the test at this point.

**Review by jbowens — CHANGES_REQUESTED** [review]
> Reviewed 2 of 8 files at r5, 1 of 6 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3044)* status: 3 of 18 files reviewed, 14 unresolved discussions (waiting on @itsbilal)

___
*[`db.go` line 1058 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGagWW982jXN-5U0Mi:-NkGagWW982jXN-5U0Mj:b-oelt7r) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/db.go#L1058)):*
> ```Go
> 		if internalOpts.snapshot.vers == nil {
> 			if internalOpts.snapshot.readState != nil {
> 				readState = internalOpts.snapshot.readState
> ```

Is it necessary to ref the `readState` because it'll be unrefed when the iterator is closed?
___
*[`db.go` line 1291 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGb1-z15tRdsAT5qw0:-NkGb1-z15tRdsAT5qw1:b-89a4l7) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/db.go#L1291)):*
> ```Go
> 	if sOpts.vers == nil {
> 		if sOpts.readState != nil {
> 			readState = sOpts.readState
> ```

Ditto
___
*[`snapshot.go` line 438 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGc86xAYmMVmYrrYTL:-NkGc86xAYmMVmYrrYTM:b-3q6ten) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/snapshot.go#L438)):*
> ```Go
> 		if err := es.mu.snap.closeLocked(); err != nil {
> 			return err
> 		}
> ```

nit: unrelated to this change, but `closeLocked` appears to be infallible. We should remove the error return value as it would make this easier to see that we're not leaking snapshot refs in an error case.
___
*[`snapshot.go` line 508 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGdHOOEmpcGZQg3Wfx:-NkGdHOOEmpcGZQg3Wfy:b-qmdeux) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/snapshot.go#L508)):*
> ```Go
> 		}
> 		sOpts.readState = es.mu.rs
> 	}
> ```

Is there a race here when `es.alwaysCreateIters == true` and `es.excised.Load() == false`? The span may be excised after we load the atomic but before we acquire the current read state.
___
*[`snapshot.go` line 559 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGeadl70kfLGpR5uNc:-NkGeadl70kfLGpR5uNd:b-uuouo2) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/snapshot.go#L559)):*
> ```Go
> 			}
> 		}
> 	}
> ```

Same race concern here too
___
*[`metamorphic/generator.go` line 1238 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGep5d-_hRp_MPfz-Y:-NkGep5d-_hRp_MPfz-Z:btdl0cj) ([raw file](https://github.com/cockroachdb/pebble/blob/d228f9df98d596dfab16a5777efb4e5868eae2f7/metamorphic/generator.go#L1238)):*
> ```Go
> 	// If we're in multi-instance mode, we must always create bounds, as we will
> 	// always create EventuallyFileOnlySnapshots to allow commands that use excises
> 	// (eg. replicateOp) to work.
> ```

I'm wondering if we should make this conditional on the op distribution (`writerIngestAndExcise > 0`), and then add a separate op distribution for `writerIngestAndExcise > 0`. It'd be nice to maintain test coverage ensuring equivalence between EFOS and ordinary snapshots in some test runs.
___
*[`metamorphic/generator.go` line 1451 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGf_FHBI1EyVa5HXpx:-NkGf_FHBI1EyVa5HXpy:b58378l) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/generator.go#L1451)):*
> ```Go
> 
> 	dbID := g.dbs.rand(g.rng)
> 	// Ingest between 1 and 3 batches.
> ```

This comment is incorrect, right? We're only ingesting one batch?
___
*[`metamorphic/ops.go` line 792 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGg1XL7ILBeFYMVgy1:-NkGg1XL7ILBeFYMVgy2:b18vjoa) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L792)):*
> ```Go
> }
> 
> type ingestAndExciseOp struct {
> ```

I think we need to add this op to `keyManager.update` (see its treatment of `ingestOp`).

We should also update `opWrittenKeys` to return the two excise keys for `ingestAndExciseOp`s.
___
*[`metamorphic/ops.go` line 803 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGglLj3rqK5GPnStt6:-NkGglLj3rqK5GPnStt7:b-igqqbx) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L803)):*
> ```Go
> 	b := t.getBatch(o.batchID)
> 	t.clearObj(o.batchID)
> 	if b.Empty() || t.testOpts.Opts.Comparer.Compare(o.exciseEnd, o.exciseStart) <= 0 {
> ```

If the excise span is invalid but the batch is non-empty, the batch is discarded? Should we always generate well-formed excise spans to avoid wasted test ops?
___
*[`metamorphic/ops.go` line 817 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGhUe814gZlqXap38t:-NkGhUe814gZlqXap38u:b-ugvrzk) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L817)):*
> ```Go
> 	path, err2 := o.build(t, h, b, 0 /* i */)
> 	if err2 != nil {
> 		h.Recordf("Build(%s) // %v", o.batchID, err2)
> ```

Missing return?
___
*[`metamorphic/ops.go` line 1445 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGhwMmC8Gd6K05fFlb:-NkGhwMmC8Gd6K05fFlc:b-j6ttla) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L1445)):*
> ```Go
> 			if err != nil {
> 				h.Recordf("%s // %v", o, err)
> 				h.history.err.Store(errors.Wrap(err, "newSnapshotOp"))
> ```

Should we panic here? The test isn't going to be able to make progress without the snapshot, and I think it'll nil pointer on the first operation that uses the snapshot.
___
*[`metamorphic/ops.go` line 1462 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGiOUXBx5QVK3PsoNU:-NkGiOUXBx5QVK3PsoNV:b-pa12o3) ([raw file](https://github.com/cockroachdb/pebble/blob/d228f9df98d596dfab16a5777efb4e5868eae2f7/metamorphic/ops.go#L1462)):*
> ```Go
> 			}
> 			return rangekey.Encode(&s, func(k base.InternalKey, v []byte) error {
> 				return w.AddRangeKey(base.MakeInternalKey(k.UserKey, 0, k.Kind()), v)
> ```

Did we not need the sequence number zeroing?
___
*[`metamorphic/ops.go` line 1535 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGiguN-lxC5GKLlpER:-NkGiguN-lxC5GKLlpES:b-7fkn9n) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L1535)):*
> ```Go
> 	var sharedSSTs []pebble.SharedSSTMeta
> 	var err error
> 	sstEmpty := true
> ```

I think we could use [`(*sstable.Writer).Metadata`](https://pkg.go.dev/github.com/cockroachdb/pebble@v0.0.0-20231122205016-4cbc644681f2/sstable#Writer.Metadata) instead to observe how many keys were written, if any.
___
*[`metamorphic/options.go` line 605 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGjReoESrjZJWkFWk1:-NkGjReoESrjZJWkFWk2:b-5m8gdc) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/options.go#L605)):*
> ```Go
> 	testOpts.useExcise = rng.Intn(2) == 0
> 	if testOpts.useExcise || testOpts.useSharedReplicate {
> 		//testOpts.efosAlwaysCreatesIters = rng.Intn(2) == 0
> ```

debugging detritus?

**Review by itsbilal — COMMENTED** [review]
TFTR! Also fixed the failure identified above. It was because `IngestAndExciseOp` had slightly divergent behaviour with `useExcise` and the only content of the batch being a rangedel that starts/ends at the same key. Moved to a better `NumEntries` based approach for identifying the no-op case in both `replicateOp` and `ingestAndExciseOp`..

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3044#-:-NkHMySJD-WOKACXe3hv:b97yy5l)* status: 3 of 18 files reviewed, 14 unresolved discussions (waiting on @jbowens)

___
*[`db.go` line 1058 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGagWW982jXN-5U0Mi:-NkH8YTTDBT0TW_y87A8:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/db.go#L1058)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Is it necessary to ref the `readState` because it'll be unrefed when the iterator is closed?
</blockquote></details>

Done.
___
*[`db.go` line 1291 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGb1-z15tRdsAT5qw0:-NkH8n-R8D-MBYewx9UB:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/db.go#L1291)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.
___
*[`snapshot.go` line 438 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGc86xAYmMVmYrrYTL:-NkH8yE3ELlwZdFilEY3:b-lttgb4) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/snapshot.go#L438)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: unrelated to this change, but `closeLocked` appears to be infallible. We should remove the error return value as it would make this easier to see that we're not leaking snapshot refs in an error case.
</blockquote></details>

That's a good point. And it looks like `snap.Close()` is similar because it returns `closeLocked`. I can do this in a separate PR.
___
*[`snapshot.go` line 508 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGdHOOEmpcGZQg3Wfx:-NkHIfIoDmsJ6PC_bIY5:bkucmit) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/snapshot.go#L508)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Is there a race here when `es.alwaysCreateIters == true` and `es.excised.Load() == false`? The span may be excised after we load the atomic but before we acquire the current read state.
</blockquote></details>

Good catch. Wasn't tripping up in the metamorphic testing because of the way it syncs operations, but fixed (by grabbing the db mutex conditionally in this method if `alwaysCreateIters` is true).
___
*[`snapshot.go` line 559 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGeadl70kfLGpR5uNc:-NkHLf6E2ELE2d1cqxbV:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/snapshot.go#L559)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Same race concern here too
</blockquote></details>

Done.
___
*[`metamorphic/generator.go` line 1238 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGep5d-_hRp_MPfz-Y:-NkHLqr-7Kx4QA_H0g3O:bo7u00l) ([raw file](https://github.com/cockroachdb/pebble/blob/d228f9df98d596dfab16a5777efb4e5868eae2f7/metamorphic/generator.go#L1238)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I'm wondering if we should make this conditional on the op distribution (`writerIngestAndExcise > 0`), and then add a separate op distribution for `writerIngestAndExcise > 0`. It'd be nice to maintain test coverage ensuring equivalence between EFOS and ordinary snapshots in some test runs.
</blockquote></details>

We still get that equivalence, right? We don't create an EFOS just because bounds are present. We just generate bounds in all cases so depending on whether the test options necessitate an EFOS (and currently none of the standard options do) or the outcome of the fibonacci hash, we could still create a regular snapshot. 
___
*[`metamorphic/generator.go` line 1451 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGf_FHBI1EyVa5HXpx:-NkH9od-9D_hps7F_mse:b-eoqoem) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/generator.go#L1451)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

This comment is incorrect, right? We're only ingesting one batch?
</blockquote></details>

Copy-paste artifact. Removed.
___
*[`metamorphic/ops.go` line 792 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGg1XL7ILBeFYMVgy1:-NkHBm3SAqA6gXO0LiJP:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L792)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we need to add this op to `keyManager.update` (see its treatment of `ingestOp`).

We should also update `opWrittenKeys` to return the two excise keys for `ingestAndExciseOp`s.
</blockquote></details>

Done.
___
*[`metamorphic/ops.go` line 803 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGglLj3rqK5GPnStt6:-NkHBRl-BAuy7wur3hpp:boj1qgt) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L803)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

If the excise span is invalid but the batch is non-empty, the batch is discarded? Should we always generate well-formed excise spans to avoid wasted test ops?
</blockquote></details>

Good catch. Done.
___
*[`metamorphic/ops.go` line 817 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGhUe814gZlqXap38t:-NkHA2o55lKKUJC-euM2:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L817)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Missing return?
</blockquote></details>

Done.
___
*[`metamorphic/ops.go` line 1445 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGhwMmC8Gd6K05fFlb:-NkHAEhaDIsSqwovFfSd:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L1445)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Should we panic here? The test isn't going to be able to make progress without the snapshot, and I think it'll nil pointer on the first operation that uses the snapshot.
</blockquote></details>

Done.
___
*[`metamorphic/ops.go` line 1462 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGiOUXBx5QVK3PsoNU:-NkHAiKO3EFxmVnpcgon:b-c48e9s) ([raw file](https://github.com/cockroachdb/pebble/blob/d228f9df98d596dfab16a5777efb4e5868eae2f7/metamorphic/ops.go#L1462)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Did we not need the sequence number zeroing?
</blockquote></details>

`ScanInternal` does it for us (and also does the appropriate sort).
___
*[`metamorphic/ops.go` line 1535 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGiguN-lxC5GKLlpER:-NkHBFtS5EOUsWbQRHQ2:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/ops.go#L1535)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we could use [`(*sstable.Writer).Metadata`](https://pkg.go.dev/github.com/cockroachdb/pebble@v0.0.0-20231122205016-4cbc644681f2/sstable#Writer.Metadata) instead to observe how many keys were written, if any.
</blockquote></details>

Done.
___
*[`metamorphic/options.go` line 605 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGjReoESrjZJWkFWk1:-NkHBNo2BX8Zs8O7bO1w:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/681a4e1070365107769894c0bedf45bab54d3761/metamorphic/options.go#L605)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

debugging detritus?
</blockquote></details>

Done.

**Review by jbowens — APPROVED** [review]
:lgtm:

> Reviewed 3 of 6 files at r8, 5 of 6 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3044#-:-NkMG8Zg-Fe_ak0fM8nU:bnfp4nl)* status: 10 of 19 files reviewed, 6 unresolved discussions (waiting on @itsbilal)

___
*[`snapshot.go` line 590 at r9](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkMCUe_1rZsqjnT-4R0:-NkMCUe_1rZsqjnT-4R1:b-mi3m25) ([raw file](https://github.com/cockroachdb/pebble/blob/4f718f2b13ed0f576acae06a8944c470091bbc7d/snapshot.go#L590)):*
> ```Go
> 	if es.alwaysCreateIters {
> 		// See the similar conditional above where we grab this mutex.
> 		es.db.mu.Unlock()
> ```

do you have any ideas on how we could clean this up? The conditional mutex acquisition is hard to follow. At the very least we could move the `opts :=` statement up above the lock acquisition to reduce the volume of code that sits within the critical section, making it a little easier to parse.

maybe we should split the entire remaining critical section into separate branches based on `alwaysCreateIters` versus `!alwaysCrateIters`? This applies to `NewIterWithContext` too I think. 
___
*[`metamorphic/generator.go` line 1238 at r8](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkGep5d-_hRp_MPfz-Y:-NkMDH_R5ub_ToLXjSTs:bkzbmgp) ([raw file](https://github.com/cockroachdb/pebble/blob/d228f9df98d596dfab16a5777efb4e5868eae2f7/metamorphic/generator.go#L1238)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

We still get that equivalence, right? We don't create an EFOS just because bounds are present. We just generate bounds in all cases so depending on whether the test options necessitate an EFOS (and currently none of the standard options do) or the outcome of the fibonacci hash, we could still create a regular snapshot.
</blockquote></details>

Ah, yeah, I wasn't reading carefully.
___
*[`metamorphic/generator.go` line 1456 at r9](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkMDrq-5oK9K5c3WtDM:-NkMDrq-5oK9K5c3WtDN:bgn4mgt) ([raw file](https://github.com/cockroachdb/pebble/blob/4f718f2b13ed0f576acae06a8944c470091bbc7d/metamorphic/generator.go#L1456)):*
> ```Go
> 	start := g.randKeyToWrite(0.001)
> 	end := g.randKeyToWrite(0.001)
> 	for g.cmp(start, end) == 0 {
> ```

nit: `g.equal(start, end)`

**Review by itsbilal — COMMENTED** [review]
TFTR!

> Dismissed @jbowens from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3044#-:-NkReyNSBT5G_fgbrAed:b-814g8k)* status: 6 of 19 files reviewed, 4 unresolved discussions (waiting on @jbowens)

___
*[`snapshot.go` line 590 at r9](https://reviewable.io/reviews/cockroachdb/pebble/3044#-NkMCUe_1rZsqjnT-4R0:-NkRezTL2JhTAqsJz63W:bratzgk) ([raw file](https://github.com/cockroachdb/pebble/blob/4f718f2b13ed0f576acae06a8944c470091bbc7d/snapshot.go#L590)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

do you have any ideas on how we could clean this up? The conditional mutex acquisition is hard to follow. At the very least we could move the `opts :=` statement up above the lock acquisition to reduce the volume of code that sits within the critical section, making it a little easier to parse.

maybe we should split the entire remaining critical section into separate branches based on `alwaysCreateIters` versus `!alwaysCrateIters`? This applies to `NewIterWithContext` too I think.
</blockquote></details>

Done. I did a separate function to handle the `alwaysCreateIters` case of `NewIterWithContext` as that function was filled with lots of conditionals on that boolean. For `ScanInternal` I just moved the opts up as it's (imo) clearer what's going on in the critical section there.
